nx/lib/nx/shared.ex

```diff
@@ -501,14 +501,14 @@ defmodule Nx.Shared do
 
   The given body is used as the default implementation otherwise.
   """
-  def default_implementation(function_name, args, default_impl)
+  def optional(function_name, args, default_impl)
       when is_atom(function_name) and is_list(args) and is_function(default_impl) do
     arity = length(args)
     backend = list_impl!(args)
 
     cond do
       backend == Nx.Defn.Expr ->
-        apply(default_impl, args)
+        Nx.Defn.Expr.optional(apply(default_impl, args), function_name, args)
```

Comment:
I would like to check, either here or inside optional, that the result of `apply(default_impl, args)` matches the output shape in `hd(args)`. We should likely check it in the third clause too.